MudTabs: Improve accessibility

nccadman19 opened the pull request:
fix(`MudTabs`): improve accessibility for keyboard interaction

- Makes tabs focusable using Arrow keys and responsive to Enter/Space keys
- Prevents key navigation into disabled tab content
- Aligns behaviour with accessibility standards (WCAG)
- Adds proper ARIA attributes for screen reader support
- Identify.Create used for the componentId
- [Label Guidance](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/examples/tabs-manual/)

Tested locally in `MudBlazor.Docs` to confirm behaviour across input methods.

## Description

This PR enhances keyboard accessibility in `MudTabs` by implementing proper keyboard navigation and ARIA support. Previously, the tabs component lacked proper keyboard interaction patterns, making it difficult for keyboard and screen reader users to navigate between tabs effectively.

**Changes include:**
- **Keyboard Navigation**: Added support for arrow key navigation (Left/Right for horizontal tabs, Up/Down for vertical tabs)
- **Activation Keys**: Tabs can now be activated using Enter and Space keys
- **Focus Management**: Proper tab index management ensures only the active tab is in the tab order
- **Disabled Tab Handling**: Keyboard navigation correctly skips disabled tabs and wraps around to enabled tabs
- **ARIA Support**: Added proper `role="tab"`, `role="tablist"`, `role="tabpanel"`, and aria-selected attributes
- **Accessibility Compliance**: Follows W3C WAI-ARIA authoring practices for tab components

## New Changes 

- Made TabsDragAndDrop_With_FiresOnItemDroppedAsync() an async task to resolve error
- Tabindex removed from 
aria-controls
- Role, id, aria-selected, tabindex and onkeydown added to mudtabs
- Indexing on the tabpanel updated to the fieldid
- Public strings made internal
- Space tests added to tabtests, keeptabsnotalive test updated to pass
- Updated gettablistid to be created oninitialized to avoid creating a new string object every time it's called
- Implemented the Delete shortcut for deletable tabs in muddynamictabs
- Added key interceptor to tab bar to prevent page scroll on space buttons

## How Has This Been Tested?

Tested locally by running the MudBlazor docs site using the edited source code:

- **Keyboard Navigation**: Verified arrow key navigation works in both horizontal and vertical tab orientations
- **Activation**: Confirmed tabs activate properly with Enter and Space keys
- **Focus Management**: Tested that focus moves correctly between enabled tabs and skips disabled ones
- **Wrapping**: Verified navigation wraps from first to last tab and vice versa
- **Screen Reader**: Tested with screen reader to confirm ARIA attributes work correctly
- **Unit Tests**: Added test coverage for keyboard interactions

## Type of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

## Checklist

- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests (manual/visual verification in Docs).

## Related Issues
- Closes #9934, #6375

nccadman19 committed:
Updated MudTabs to align with W3C ARIA tabs pattern

- added role="tab", role="tabpanel" and role="tablist" attributes
- implemented keyboard navigation (ArrowLeft, ArrowRight, Enter, Space) with manual activation
- added tabindex handling for correct focus behaviour
- added unit tests to validate keyboard navigation, attributes, and focus behaviour

nccadman19 committed:
Merge branch 'dev' into fix/accessible-tabs

nccadman19 committed:
Updated to spaces

nccadman19 committed:
Summary added to tests

nccadman19 committed:
Removed redundant test

nccadman19 committed:
roles updated to reflect wcag standards

nccadman19 committed:
Updated KeepTabs Not AliveTest to deal with div inside MudTabPanel

danielchalmers commented on src/MudBlazor/Components/Tabs/MudTabPanel.razor:
```diff
@@ -4,14 +4,16 @@
     @if (Parent?.ActivePanel == this)
     {
-        @ChildContent
+        <div role="tabpanel" tabindex="0">
```
Needs a class name and I'd imagine `display: contents` but...

I'd love if someone with more knowledge of the tab component than me could answer why this wasn't inside a `div` before, or shouldn't have `@attributes="UserAttributes" class="@Classname" style="@Stylename"` like the container above it?

versile2 replied:
It should definitely match what's above it. "mud-tab-panel" at minimum, but I think just using the `@Class` will work, but it should match the other exactly.

nccadman19 replied:
Okay added to the most recent push. 

```
@if (Parent?.KeepPanelsAlive == true)
{
	<div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" tabindex="0" id="@Parent.GetPanelId(this)" aria-labelledby="@Parent.GetTabId(this)">
        @ChildContent
    </div>
}
else
{
    @if (Parent?.ActivePanel == this)
    {
		<div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" tabindex="0" id="@Parent.GetPanelId(this)" aria-labelledby="@Parent.GetTabId(this)">
            @ChildContent
        </div>
    }
}
```

ar-schraml replied:
@nccadman19 Sorry for commenting on an old PR. But there's an unresolved issue here, and I'm currently not sure if I should just create a new issue for it.

`Stylename` applies `display: contents` only on the condition that `KeepPanelsAlive` is true. See https://github.com/MudBlazor/MudBlazor/blob/b04d0e9cde2f8e10ddf947efa44870ee2e0bdc78/src/MudBlazor/Components/Tabs/MudTabPanel.razor.cs#L19-L23

Meaning that `display: contents` is missing without `KeepPanelsAlive`. So even though the CSS classes and styles seem to be identical with `KeepPanelsAlive` and without it, they actually aren't.

In the [PR](https://github.com/MudBlazor/MudBlazor/pull/10797) that added this condition there even was [this comment by @ScarletKuro](https://github.com/MudBlazor/MudBlazor/pull/10797/changes#r1938895863)
> @danielchalmers should we move this to the class?
Also, do you think `Parent?.KeepPanelsAlive == true` is necessary considering that it's already insane a similar if?

With @danielchalmers replying:
> Yeah moving it to `Classname` would be ideal and it seems like the condition can be removed since that render path is only used in the same condition anyway.

But unfortunately the condition wasn't removed.

All of this leads to an issue I'm currently facing where I made my `MudTabs` a flex box and applied some flex styles to the `MudTabPanel`s via `MudTabs`'s `PanelClass` property (renamed to `TabPanelsClass` in #12156). But because this `<div>` without `display: contents` sits between them, the flex behaviour doesn't work as expected.
It does work as soon as I add `KeepPanelsAlive="true"` though, as then `display: contents` gets applied to the `<div>` as described above.
I hope I was able to explain that in a way that makes sense.

I think the solution to this should be as easy as to remove the `Parent?.KeepPanelsAlive == true` condition on `Stylename`.

So should I create a new issue for this?

danielchalmers reviewed (commented):
I need to test locally but always love to see more accessibility!

There's a number of formatting diffs, could you see if you can revert them so it's easier to review the real changes?

nccadman19 committed:
Formatting updated to match old code

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -1090,5 +1090,103 @@ internal void ItemUpdated(MudItemDropInfo<MudTabPanel> dropItem)
+                        handled = true;
+                    }
+                    break;
+            }
```
Removed handled flag, unsure if a e.PreventDefault is required but have removed it for now.

```
        private async Task HandleTabKeyDownAsync(KeyboardEventArgs e, MudTabPanel panel)
        {
            switch (e.Key)
            {
                case "Enter":
                case " ":
                    ActivatePanel(panel, null, false);
                    break;

                case "ArrowLeft":
                    if (!IsVerticalTabs())
                    {
                        await MoveFocusToPreviousTab(panel);
                    }
                    break;
```

nccadman19 committed:
Remove await is handled, not required

nccadman19 committed:
feat: Add ARIA compliance to MudTabs component

- Add stable ID generation for tabs and panels
- Implement aria-controls and aria-labelledby attributes
- Use panel index for indentification
- Ensure bidirectional accessibility relationship between tabs and panels

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabs.razor:
```diff
@@ -100,39 +100,41 @@
-    RenderFragment RenderTab(MudTabPanel panel) => @<div @ref="panel.PanelRef" class="@GetTabClass(panel)" style="@GetTabStyle(panel)" @onclick=@( e => ActivatePanel(panel, e, false) )>
+    RenderFragment RenderTab(MudTabPanel panel) => 
+    @<div @ref="panel.PanelRef" class="@GetTabClass(panel)" style="@GetTabStyle(panel)" @onclick=@( e => ActivatePanel(panel, e, false) ) 
+        tabindex="@(panel.Disabled ? "-1" : ActivePanel == panel ? "0" : "-1")" @onkeydown="(e => HandleTabKeyDownAsync(e, panel))" role="tab" aria-selected="@(ActivePanel == panel ? "true" : "false")">
```
Addressed in latest commit

```
        /// <summary>
        /// Generates a unique ID for a tab element using the panel's index.
        /// Required for aria-controls attribute to link tab to its panel.
        /// </summary>
        public string GetTabId(MudTabPanel panel)
        {
            var panelIndex = _panels.IndexOf(panel);
            return $"mud-tabs-{_componentId}-tab-{panelIndex}";
        }

        /// <summary>
        /// Generates a unique ID for a panel element using the panel's index.
        /// Required for aria-labelledby attribute to link panel back to its tab.
        /// </summary>
        public string GetPanelId(MudTabPanel panel)
        {
            var panelIndex = _panels.IndexOf(panel);
            return $"mud-tabs-{_componentId}-panel-{panelIndex}";
        }
```

   ```
 RenderFragment RenderTab(MudTabPanel panel) => @<div @ref="panel.PanelRef" class="@GetTabClass(panel)" style="@GetTabStyle(panel)" @onclick=@( e => ActivatePanel(panel, e, false) ) 
        tabindex="@(panel.Disabled ? "-1" : ActivePanel == panel ? "0" : "-1")" @onkeydown="(e => HandleTabKeyDownAsync(e, panel))" role="tab" aria-selected="@(ActivePanel == panel ? "true" : "false")"
		id="@GetTabId(panel)" aria-controls="@GetPanelId(panel)">
```

nccadman19 committed:
Merge branch 'dev' into fix/accessible-tabs

versile2 reviewed (changes requested):
Apart from the MudTabPanel needing the class/style/attributes this feels ready to go.

nccadman19 committed:
Added attributes, class and style support to tab panels

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabPanel.razor:
```diff
@@ -4,14 +4,16 @@
 @if (Parent?.KeepPanelsAlive == true)
 {
-    <div @attributes="UserAttributes" class="@Classname" style="@Stylename">
+    <div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" tabindex="0">
```
Added to most recent push.

```
@if (Parent?.KeepPanelsAlive == true)
{
	<div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" tabindex="0" id="@Parent.GetPanelId(this)" aria-labelledby="@Parent.GetTabId(this)">
        @ChildContent
    </div>
}
else
{
    @if (Parent?.ActivePanel == this)
    {
		<div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" tabindex="0" id="@Parent.GetPanelId(this)" aria-labelledby="@Parent.GetTabId(this)">
            @ChildContent
        </div>
    }
}
```

versile2 commented on src/MudBlazor/Components/Tabs/MudTabs.razor:
```diff
@@ -21,9 +21,9 @@
                         @if (EnableDragAndDrop)
                         {
-                            <MudDropContainer T="MudTabPanel" Class="@WrapperClassnames" Style="@WrapperScrollStyle" Items="_panels" 
+                            <MudDropContainer T="MudTabPanel" Class="@WrapperClassnames" Style="@WrapperScrollStyle" Items="_panels"
```
Did you verify the role="tablist" works properly in EnableDragandDrop mode?

nccadman19 replied:
On the drag and drop mode the panels are index 0-8 in that order regardless of the drag and drop. I'm assuming that's the preferred outcome, so you can drag the panels in any order and the panel index reorders to that if nine has been dragged to the third position it would say panel-2 not panel-8 is that what you were expecting?

```
<div role="tablist" style="transform:translateY(-0px);" class="mud-drop-container mud-tabs-tabbar-wrapper mud-tabs-vertical">
  <div class="mud-tab mud-ripple  mud-tab-panel" style="min-width:160px;" tabindex="-1" role="tab" aria-selected="false" id="mud-tabs-15bfde4f-tab-0" aria-controls="mud-tabs-15bfde4f-panel-0" _bl_5d8e2f97-bcd8-4a0f-ac66-304b3fc3eb17="">Two</div>
  <div class="mud-tab mud-ripple  mud-tab-panel" style="min-width:160px;" tabindex="-1" role="tab" aria-selected="false" id="mud-tabs-15bfde4f-tab-1" aria-controls="mud-tabs-15bfde4f-panel-1" _bl_21383065-aeb8-498b-a5bd-19db3f4353d2="">Six</div>
  <div class="mud-tab mud-tab-active mud-ripple   mud-tab-panel" style="min-width:160px;" tabindex="0" role="tab" aria-selected="true" id="mud-tabs-15bfde4f-tab-2" aria-controls="mud-tabs-15bfde4f-panel-2" _bl_b5b2fd61-42ae-4eda-b0a3-0c89de27f155="">Three</div>
  <div class="mud-tab mud-ripple  mud-tab-panel" style="min-width:160px;" tabindex="-1" role="tab" aria-selected="false" id="mud-tabs-15bfde4f-tab-3" aria-controls="mud-tabs-15bfde4f-panel-3" _bl_94e4f37c-8007-43a3-bf68-e5dc8dd56cb2="">Four</div>
  <div class="mud-tab mud-ripple  mud-tab-panel" style="min-width:160px;" tabindex="-1" role="tab" aria-selected="false" id="mud-tabs-15bfde4f-tab-4" aria-controls="mud-tabs-15bfde4f-panel-4" _bl_bbe21be4-f8de-4100-a35a-3a1a4ea6fcb6="">Nine</div
  <div class="mud-tab mud-ripple  mud-tab-panel" style="min-width:160px;" tabindex="-1" role="tab" aria-selected="false" id="mud-tabs-15bfde4f-tab-5" aria-controls="mud-tabs-15bfde4f-panel-5" _bl_8518752b-c87d-4bff-936a-55e3841d63ab="">Five</div>
  <div class="mud-tab mud-ripple  mud-tab-panel" style="min-width:160px;" tabindex="-1" role="tab" aria-selected="false" id="mud-tabs-15bfde4f-tab-6" aria-controls="mud-tabs-15bfde4f-panel-6" _bl_e3ad4668-2ebc-454c-b4b0-75800f4e85ed="">One</div>
</div>
```

ScarletKuro commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -452,6 +452,13 @@ public int ActivePanelIndex
+        /// Used to generate stable, unique IDs for tabs and panels to ensure ARIA compliance.
+        /// Prevents ID conflicts when multiple tab components exist on the same page.
+        /// </summary>
+        private readonly string _componentId = Guid.NewGuid().ToString("N")[..8];
```
We have `Identifier.Create` API for this.

ScarletKuro commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -1090,5 +1097,110 @@ internal void ItemUpdated(MudItemDropInfo<MudTabPanel> dropItem)
+        /// Generates a unique ID for a tab element using the panel's index.
+        /// Required for aria-controls attribute to link tab to its panel.
+        /// </summary>
+        public string GetTabId(MudTabPanel panel)
```
I think I'd make this API and `GetPanelId` to be internal for now.
@igotinfected @danielchalmers Do we need at all the panel index information for the id `id="@..."`? Why can't we just return straight the `_componentId` which will be unique anyway, I don't think the index does much here.

danielchalmers replied:
For aria you only need a unique ID for each tab. Agree on internal/private

nccadman19 replied:
Hey, I was following this example: [Example of Tabs with Manual Activation](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/examples/tabs-manual/)
I have kept it public so the `MudTabPanel.razor` can reach the `id` and `aria-labelledby` but I have updated the list to only store `tablist-id`.

Here's the code example I am following, in the `tabpanel` section it uses a `id="tabpanel-1"` and a `aria-labelledby="tab-1"`. Which I have updated in the most recent push with the `tablist-{_componentId}` at the beginning so the user can differentiate between different `tablist` in case there is more than one on the page. 

```
<div class="tabs">
  <h3 id="tablist-1">Danish Composers</h3>
  <div role="tablist" aria-labelledby="tablist-1" class="manual">
    <button id="tab-1" type="button" role="tab" aria-selected="true" aria-controls="tabpanel-1">
      <span class="focus">Maria Ahlefeldt</span>
    </button>
    <button id="tab-2" type="button" role="tab" aria-selected="false" aria-controls="tabpanel-2" tabindex="-1">
      <span class="focus">Carl Andersen</span>
    </button>
    <button id="tab-3" type="button" role="tab" aria-selected="false" aria-controls="tabpanel-3" tabindex="-1">
      <span class="focus">Ida da Fonseca</span>
    </button>
    <button id="tab-4" type="button" role="tab" aria-selected="false" aria-controls="tabpanel-4" tabindex="-1">
      <span class="focus">Peter Müller</span>
    </button>
  </div>

  <div id="tabpanel-1" role="tabpanel" aria-labelledby="tab-1">
    <p>
      <a href="https://en.wikipedia.org/wiki/Maria_Theresia_Ahlefeldt">Maria Theresia Ahlefeldt</a>
      (16 January 1755 – 20 December 1810) was a Danish, (originally German), composer.
      She is known as the first female composer in Denmark.
      Maria Theresia composed music for several ballets, operas, and plays of the royal theatre.
      She was given good critic as a composer and described as a “<span lang="da">virkelig Tonekunstnerinde</span>” ('a True Artist of Music').
    </p>
  </div>
  <div id="tabpanel-2" role="tabpanel" aria-labelledby="tab-2" class="is-hidden">
    <p>
      <a href="https://en.wikipedia.org/wiki/Joachim_Andersen_(composer)">Carl Joachim Andersen</a>
      (29 April 1847 – 7 May 1909) was a Danish flutist, conductor and composer born in Copenhagen, son of the flutist Christian Joachim Andersen.
      Both as a virtuoso and as composer of flute music, he is considered one of the best of his time.
      He was considered to be a tough leader and teacher and demanded as such a lot from his orchestras but through that style he reached a high level.
    </p>
  </div>
  <div id="tabpanel-3" role="tabpanel" aria-labelledby="tab-3" class="is-hidden">
    <p>
      <a href="https://en.wikipedia.org/wiki/Ida_Henriette_da_Fonseca">Ida Henriette da Fonseca</a>
      (July 27, 1802 – July 6, 1858) was a Danish opera singer and composer.
      Ida Henriette da Fonseca was the daughter of Abraham da Fonseca (1776–1849) and Marie Sofie Kiærskou (1784–1863).
      She and her sister Emilie da Fonseca were students of Giuseppe Siboni, choir master of the Opera in Copenhagen.
      She was given a place at the royal Opera alongside her sister the same year she debuted in 1827.
    </p>
  </div>
  <div id="tabpanel-4" role="tabpanel" aria-labelledby="tab-4" class="is-hidden">
    <p>
      <a href="https://en.wikipedia.org/wiki/Peter_Lange-M%C3%BCller">Peter Erasmus Lange-Müller</a>
      (1 December 1850 – 26 February 1926) was a Danish composer and pianist.
      His compositional style was influenced by Danish folk music and by the work of Robert Schumann; Johannes Brahms; and his Danish countrymen, including J.P.E. Hartmann.
    </p>
  </div>
</div>
```

I am on leave from **4pm BST 05/07/2025 - 21/07/2025** so I wont be able to make any changes until after I am back!

igotinfected replied:
Unique ids for tabs and tabpanels are indeed necessary for the `aria-controls`/`aria-labelledby`.

@nccadman19 The `MudTabPanel` will still be able to call these methods if you change it to `internal`, the advantage is we avoid exposing methods to the public and causing breaking changes down the line if we ever need to change them. You could also give the `MudTabPanel` an internal id if it doesn't have one already and use that instead to avoid the extra `IndexOf` and new strings.

nccadman19 committed:
Updated mudtab aria labels to match: https://www.w3.org/WAI/ARIA/apg/patterns/tabs/examples/tabs-manual/

igotinfected commented on src/MudBlazor.UnitTests/Components/TabsTests.cs:
```diff
@@ -125,7 +126,7 @@ public void KeepTabsAliveTest()
 
         /// <summary>
-        /// When KeepPanelsAlive="true" the panels are not destroyed and recreated on tab-switch. We prove that by using a button click counter on every tab and
+        /// When KeepPanelsAlive="false" the panels are not destroyed and recreated on tab-switch. We prove that by using a button click counter on every tab and
```
I believe the previous `KeepPanelsAlive="true"` was correct? `KeepPanelsAlive="true"` keeps the panels alive, avoiding recreating them when switching between tabs

igotinfected commented on src/MudBlazor.UnitTests/Components/TabsTests.cs:
```diff
@@ -1478,5 +1485,83 @@ public void LabelSorting_CustomSortComparer()
+        /// Tab selection changes on keyboard Up and Down arrow keys, is activated by Enter/Space keys
+        /// </summary>
+        [Test]
+        public async Task VerticalTabs_SupportsArrowUpDownNavigation()
```
Would add a test (or an some extra lines in the existing tests) triggering a tab selection with `Space` to protect against that breaking accidentally

igotinfected commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -1090,5 +1097,119 @@ internal void ItemUpdated(MudItemDropInfo<MudTabPanel> dropItem)
+        /// Generates a unique ID for a tab list. 
+        /// Required for aria-controls attribute to identify each tablist.
+        /// </summary>
+        public string GetTabListId()
```
Could create this once in constructor/oninitialized to avoid creating a new string object every time it's called

igotinfected reviewed (changes requested):
Appreciate your accessibility contribution!!

Added a few comments and tested the changes locally, some feedback:

- in the "scrollable list" example the scroll icons aren't updated when hitting the last page/item ("go right" button should disappear, "go left" button should appear) and sometimes the view doesn't scroll with the focus, such that the focused tab is out of view ([local WASM docs test URL](https://localhost:5001/components/tabs#scrolling-tabs))
- in the "visibility" example, toggling the visibility of the second tab breaks keyboard navigation (switching focus between the visible tabs doesn't work as expected) ([local WASM docs test URL](https://localhost:5001/components/tabs#visibility))
- if you're up for it you could also implement the `Delete` shortcut for deletable tabs in [this local WASM docs example](https://localhost:5001/components/tabs#dynamic-tabs)) for example
- in the "simple tabs" example the tab wrapper has an `aria-labelledby` that points to an element that doesn't exist, I believe this is something that should be user controlled if they provide a label for the tab component
- pressing `Space` to activate a tab makes the page jump, you can refer to [MudCheckBox](https://github.com/MudBlazor/MudBlazor/blob/7fed473fdd3bb0d7a8a60f4a7b4dd52d59cd0568/src/MudBlazor/Components/CheckBox/MudCheckBox.razor.cs#L199-L216)'s implementation for how to fix that

Don't necessarily think any of these are blocking for the PR though!

nccadman19 committed:
merge to main

nccadman19 committed:
Tab tests rerfactored, made TabsDragAndDrop_With_FiresOnItemDroppedAsync() an async task to resolve error

nccadman19 committed:
Refactored: tabindex removed from tabpanel
aria-controls, role, id, aria-selected, tabindex and onkeydown added to mudtabs
indexing on the tabpanel updated to the fieldid
public string made internal
space tests added to tabtests, keeptabsnotalive test updated to pass

nccadman19 committed:
Refactor: updated gettablistid to be created oninitialized to avoid creating a new string object every time it's called

nccadman19 committed:
Implemented the Delete shortcut for deletable tabs in muddynamictabs

nccadman19 committed:
Added key interceptor to tab bar to prevent page scroll on space buttons

igotinfected reviewed (approved):
LGTM!

For posterity, these issues are outstanding but as mentioned before, the current implementation is better than what we had before ;)

- in the "scrollable list" example the scroll icons aren't updated when hitting the last page/item ("go right" button should disappear, "go left" button should appear) and sometimes the view doesn't scroll with the focus, such that the focused tab is out of view ([local WASM docs test URL](https://localhost:5001/components/tabs#scrolling-tabs))
- in the "visibility" example, toggling the visibility of the second tab breaks keyboard navigation (switching focus between the visible tabs doesn't work as expected) ([local WASM docs test URL](https://localhost:5001/components/tabs#visibility))
- in the "simple tabs" example the tab wrapper has an aria-labelledby that points to an element that doesn't exist, I believe this is something that should be user controlled if they provide a label for the tab component

nccadman19 committed:
Aria labelled by removed

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -487,7 +500,6 @@ protected override void OnParametersSet()
         protected override async Task OnAfterRenderAsync(bool firstRender)
         {
             await base.OnAfterRenderAsync(firstRender);
```
Resolved in latest commit

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -1110,5 +1139,126 @@ internal async Task ItemUpdated(MudItemDropInfo<MudTabPanel> dropItem)
+        }
+
+        /// <summary>
+        /// Allows the user to move to the next tab using keyarrow
```
Resolved in latest commit

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabs.razor.cs:
```diff
@@ -1110,5 +1139,126 @@ internal async Task ItemUpdated(MudItemDropInfo<MudTabPanel> dropItem)
+        }
+
+        /// <summary>
+        /// Allows the user to move to the previous tab using keyarrow
```
Resolved in latest commit

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabPanel.razor:
```diff
@@ -4,14 +4,16 @@
 @if (Parent?.KeepPanelsAlive == true)
 {
-    <div @attributes="UserAttributes" class="@Classname" style="@Stylename">
+	<div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" id="@Parent.GetTabPanelId(this)" aria-labelledby="@Parent.GetTabId(this)">
```
Resolved in latest commit

nccadman19 commented on src/MudBlazor/Components/Tabs/MudTabPanel.razor:
```diff
@@ -4,14 +4,16 @@
     @if (Parent?.ActivePanel == this)
     {
-        @ChildContent
+		<div @attributes="UserAttributes" class="@Classname" style="@Stylename" role="tabpanel" id="@Parent.GetTabPanelId(this)" aria-labelledby="@Parent.GetTabId(this)">
```
Resolved in latest commit

nccadman19 commented on src/MudBlazor/Components/Tabs/MudDynamicTabs.razor:
```diff
@@ -29,15 +29,17 @@
                                        Style="@CloseIconStyle"
-                                       OnClick="() => CloseTab.InvokeAsync(context)"/>
+                                       OnClick="() => CloseTab.InvokeAsync(context)"
+									   @onkeydown="@(e => HandleTabKeyDownAsync(e, context))"/>
```
Resolved in latest commit

nccadman19 commented on src/MudBlazor/Components/Tabs/MudDynamicTabs.razor:
```diff
@@ -29,15 +29,17 @@
                                    Style="@CloseIconStyle"
-                                   OnClick="() => CloseTab.InvokeAsync(context)"/>
+                                   OnClick="() => CloseTab.InvokeAsync(context)"
+								   @onkeydown="@(e => HandleTabKeyDownAsync(e, context))"/>
```
Resolved in latest commit

nccadman19 committed:
Test added for the GetTabListId to appease codecov response
Copilot review changes applied, spacing fixed

nccadman19 committed:
Merge branch 'dev' into fix/accessible-tabs

danielchalmers reviewed (approved):
Looks great, thanks for your patience!
